Add server-side apply operation

mruoss opened the pull request:
Implements apply the same way as kubectl does - by storing a json representation of the last applied resource in an annotation.

Since the apply has to check whether the resource already exists in the cluster, it requires the run() operation to be passed as a callback. Finally, this function applies the resource to the cluster directly by calling the run callback.

resolves #13

coryodaniel commented:
Nice, this is awesome! To remain consistent w/ the API and not have to provide a callback, could you create a [runner (K8s.Client.Runner.Declarative?)](https://github.com/coryodaniel/k8s/tree/master/lib/k8s/client/runner) that uses `Base.run/N`?

mruoss commented:
Yes, that bothered me, too. Now as far as I understood the API, runners take an `%Operation{}` struct as argument. However, if I create this declarative runner and move over my code, that runner's run function would take a resource as argument, not an operation. Is that consistent enough? Or did you have a different approach in mind?

coryodaniel commented:
Very. 
Good. 
Point. 

The interaction is different here. The apply would create potentially two `Operation`s from a resource. 

Do you think it would be odd to have a set of “declarative” methods on K8s.Client that work on a resource map (or struct) Since the others all work on Operation?

I could see two nested modules of Client, Imperative and Declarative and then alias both to Client as convenience. 

The API was designed so operations could be passed to any runner, but I could easily see a use case where operator developers just want the declarative equivalent of kubectl for apply and delete.

shanesveller commented:
What are some of the motives for capturing the annotation - is it for auditing or reference purposes, for intentional compatibility with falling back to `kubectl apply` and/or other out-of-band management, or perhaps is it being inspected by other code elsewhere? Something else I'm not imagining?

If you're targeting Kubernetes 1.16+ you also have some options with [server-side application](https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply) that may satisfy some of the same goals, but not all. I think it would potentially need more intentional support in `k8s` as it includes use of a new `managedFields` key within the various resource `spec` bodies, and a change to the `Content-Type` header.

Here's a few more references to get the ball rolling:
- Proposal/RFC https://github.com/kubernetes/enhancements/issues/555
- [Surface-level definition](https://github.com/clux/kube-rs/blob/09dd6b303dc6caec7e8b0e44db241d40e7f90aea/kube/src/api/typed.rs#L193-L235) in the Rust `kube` crate 
- One of the test cases in the upstream Go code [here](https://github.com/kubernetes/kubernetes/blob/1f14cbac544968e4c9e959f31850d4e84b38ab4b/test/integration/apiserver/apply/apply_test.go#L289-L375) which for me is abstracted too far to help reverse-engineer the payload

mruoss commented:
Sorry, I was on vacation until now. `kubectl apply` stores the applied resource inside an annotation and uses it later on for change detection. My code does the same.

I know about the server-side but I am targeting support for an earlier version of Kubernetes, too. I propose to finish this pull request and put the discussion around server-side application into a separate issue - or what are your thoughts on this @coryodaniel?

shanesveller commented on lib/k8s/client/declarative.ex:
```diff
@@ -0,0 +1,90 @@
+  alias K8s.Client.Imperative
+  alias K8s.Client.Runner.Base
+
+  @last_applied_configuration_annotation "ex.kubernetes.io/last-applied-configuration"
```
This probably shouldn't use anything on `kubernetes.io` even by appearance. I'd orient it around the GitHub origin of this library or just make it configurable, with a non-`kubernetes.io` default.

Maybe something like this, I'm sure there's a limit on max length to contend with. WDYT @coryodaniel? 

```suggestion
  @last_applied_configuration_annotation "k8s.corydaniels.github.com/last-applied-configuration"
```

shanesveller replied:
🤦 Sorry about mucking up your last name, Cory 😳

shanesveller commented on lib/k8s/client/declarative.ex:
```diff
@@ -0,0 +1,90 @@
+
+  @spec add_last_applied_configuration(map) :: map
+  defp add_last_applied_configuration(resource) do
+    update_in(resource, ~w(metadata annotations), fn annotations ->
+      (annotations || %{})
+      |> Map.put(
+           @last_applied_configuration_annotation,
+           Jason.encode!(resource)
+         )
+    end)
```
```suggestion
    put_in(resource, ["metadata", Access.key("annotations", %{}), @last_applied_configuration_annotation]), Jason.encode!(resource))
    end)
```

shanesveller commented on lib/k8s/client/declarative.ex:
```diff
@@ -0,0 +1,90 @@
+  @last_applied_configuration_annotation "ex.kubernetes.io/last-applied-configuration"
+
+  @doc """
+  Like kubectl it applies the given resource to the cluster. Since the apply has to check whether the resource already
```
I would be very conservative in comparisons to `kubectl`, particularly in phrasing that reads as "this implements `kubectl`'s behavior". "Inspired by" would be preferable.

Adjusting for linebreaks, etc.
```suggestion
  Inspired by `kubectl apply`, updates a given resource on the cluster, while recording the previous state in the annotation `(#{@last_applied_configuration_annotation})`. Since the apply has to check whether the resource already
```

shanesveller commented on CHANGELOG.md:
```diff
@@ -9,13 +9,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - K8s.Operation.get_query_param/3 to get a query parameter by key
 - DigitalOcean authentication
 - `K8s.Resource.NamedList.access/1` - Accessor for lists with named items (containers, env, ...) ([#82](https://github.com/coryodaniel/k8s/pull/82))
+- `K8s.Client.apply/2` - A kubectl style apply function ([#81](https://github.com/coryodaniel/k8s/pull/81))
```
```suggestion
- `K8s.Client.apply/2` - Updates a resource while recording its previous state, inspired by `kubectl` ([#81](https://github.com/coryodaniel/k8s/pull/81))
```

shanesveller commented on lib/k8s/client.ex:
```diff
@@ -22,7 +22,7 @@ defmodule K8s.Client do
   @type options :: [option]
 
-  alias K8s.Operation
+  alias K8s.Client.{Declarative, Imperative}
```
I don't know that I'm totally sold on this dichotomy needing to exist or being desirable, or on including it in this PR. I think the new `apply` function could've lived wherever and that a refactor should've been a separate issue/PR, because now Cory has to decide whether to reject some of your changes while valuing others, and either accepting the subjective along with the good or forcing you to re-work. Part of being neighborly in OSS is working to make your PRs transparent and reviewer-friendly. Just my $0.02.

mruoss replied:
Fair enough. What do you think, @coryodaniel? How do you want to handle this?

shanesveller commented on lib/k8s/client/imperative.ex:
```diff
@@ -0,0 +1,734 @@
+defmodule K8s.Client.Imperative do
```
This file is a strict move of all of the previous content from `client.ex`, yes?

mruoss replied:
Not strict and all of the previous content, but mostly. As I wrote, here I'm not sure I understood correctly what was suggested.

shanesveller reviewed (commented):
Disclaimer: I'm not a maintainer, just an opinionated and invested bystander. None of my feedback is binding or blocking, but is well-intentioned.

alexgleason commented:
I had a similar problem and solved it with this util function:

```elixir
def get_or_create(resource, conn) do
  with {:ok, _} = result <-
         resource
         |> K8s.Client.get()
         |> K8s.Client.run(conn) do
    result
  else
    _ ->
      resource
      |> K8s.Client.create()
      |> K8s.Client.run(conn)
  end
end
```

It's slow, but works great behind a background worker like [Oban](https://github.com/sorentwo/oban).

Commit:
fix typespec for headers

Commit:
build an operation for applying resources

Commit:
set headers for PATCH in Base module

Commit:
wire up the apply operation

Commit:
update changelog

mruoss commented:
Stumbled upon this yesterday. Guess it wasn't going anywhere, so let's try this by building a server-side apply operation.

mruoss commented on lib/k8s/middleware/request/initialize.ex:
```diff
@@ -9,7 +9,7 @@ defmodule K8s.Middleware.Request.Initialize do
     with {:ok, request_options} <- K8s.Conn.RequestOptions.generate(conn) do
       request_option_headers = conn.http_provider.headers(method, request_options)
-      updated_headers = Keyword.merge(headers, request_option_headers)
+      updated_headers = headers ++ request_option_headers
```
it's either this or defining headers as keyword lists: `[{:"Content-Type", "application/json"},{:"Accept", "application/json"}]`

coryodaniel replied:
Ahh. I thought they were keyword lists. I intended to do that when I moved query parameters from Maps to keyword lists.

I wouldn’t be against it but I would need to do a minor version bump

mruoss replied:
Ahm... I didn't get which one you prefer... keyword lists or lists of binary/binary tuples?

Probably keyword lists make more sense. Keys are all generated in code...

coryodaniel replied:
I like keyword lists there, and I think they make sense, but for anyone familiar w HTTPoison (leaky abstraction aside) is it confusing to have a different syntax than b/b tuples?

mruoss replied:
I do agree. Looking at the [typespec of HTTPoison headers](https://hexdocs.pm/httpoison/HTTPoison.Request.html#t:headers/0), keyword lists should be fine, no?
I was confused that the typespec (of %Request{}) said keyword list while the code did b/b tuples. I think it would even have raised an exception if request.headers was not empty (Keyword.merge/2 does not work for b/b tuples). But I'll change this to keyword lists.

Commit:
use keyword lists for headers

Commit:
deprecate old headers/2 function

mruoss commented on lib/k8s/client/http_provider.ex:
```diff
@@ -119,6 +129,7 @@ defmodule K8s.Client.HTTPProvider do
       [{"Accept", "application/json"}, {"Content-Type", "application/json"}, {"Authorization", "Basic AF"}]
   """
   @impl true
+  @deprecated "Use headers/1 instead"
```
Instead of changing the old function, I marked it as deprecated and created a new one.

mruoss commented:
next question: Should more merge types be supported (json and strategic) - maybe in separate ticket?
More infos e.g. on https://blog.atomist.com/kubernetes-apply-replace-patch/

mruoss commented:
Hey Cory, it would be nice to get this merged soonish. Currently it is not possible for me to implement this on my side because I can't control the headers. How can we support? Shall I try to find someone to give this PR a review? Or would you be interested in co-maintainership for this and bonny?

And I guess next - because I'm missing this for an operator - I'd like to port it back to the version used by bonny...

Thanks,
Michael

coryodaniel commented:
I’d like to see an [integration test](https://github.com/coryodaniel/k8s/blob/ed12aed3049b5bacaece7d3d180100c73540e658/Makefile#L34) and a clean GH actions pass.

I’ve been slammed getting my startup off the ground. I’ve had these PRs in my todos for a Friday afternoon but keep getting deprioritized.

Could you also explain the depreciation? I recall the http method being important for that function, I think it had to do with streaming or prepping for pod-exec? Can’t recall honestly.

coryodaniel commented:
@mruoss I could definitely use some help here and on Bonny if you’re interested in being added as a maintainer, send me an email. It’s  cory@GITHUB_USERNAME.com

Commit:
add integration tests for apply

mruoss commented:
> I’d like to see an [integration test](https://github.com/coryodaniel/k8s/blob/ed12aed3049b5bacaece7d3d180100c73540e658/Makefile#L34) 

Good point! I have added two tests. They pass on my end if I run `make test.integration`. The only thing I'm not 100% sure: are they in the right place? Or would they belong to a new `operation_integration_test.ex`?

> ...and a clean GH actions pass.

OK, I had to fix some unrelated typespec for that. (see last commit)
 
> I’ve been slammed getting my startup off the ground. I’ve had these PRs in my todos for a Friday afternoon but keep getting deprioritized.

No worries, I know this really needs time. It's just that if the feedback comes so late, I'm again totally out of the code. But I totally see your point about priorities. 

> Could you also explain the depreciation? I recall the http method being important for that function, I think it had to do with streaming or prepping for pod-exec? Can’t recall honestly.

Absolutely. I deprecated `headers/2` in favor of a new `headers/1`. 
Before, the function would always set the content-type to "application/merge-patch+json" for a PATCH request. But the content type actually doesn't just depend on the HTTP method. A PATCH request can have different Content-Type headers now, depending on the type (patch, apply, etc...). So I moved the logic that defines the content-type to base.ex. `headers/1` now just sets the content-type to "application/json" if it is missing.

I could also just have changed the old function - but it is a public function on that module and that would break BC - so I kept it.

> ... send me an email

will do.

mruoss commented on test/k8s/client/runner/base_integration_test.exs:
```diff
@@ -96,6 +96,43 @@ defmodule K8s.Client.Runner.BaseIntegrationTest do
+      result = K8s.Client.run(conn, operation)
+      assert {:ok, _pod} = result
+
+      assert is_nil(pod["metadata"]["labels"]["some"])
```
move to line 114 (and add a comment)

mruoss commented on lib/k8s/client.ex:
```diff
@@ -52,9 +52,57 @@ defmodule K8s.Client do
+      }
+  """
+  @spec apply(map(), binary(), boolean()) :: Operation.t()
+  def apply(resource, field_manager \\ "elixir", force \\ true)
```
Use keyword list for `field_manager` and `force` (?)

mruoss reviewed (commented):
Review with @jlgeering

Commit:
change API (client.ex) - make management params keyword list

Commit:
add more integration tests

jlgeering commented on lib/k8s/client.ex:
```diff
@@ -78,30 +78,58 @@ defmodule K8s.Client do
+        mgmt_params \\ []
+      ) do
+    field_manager = Keyword.get(mgmt_params, :field_manager, "Elixir")
+    force = Keyword.get(mgmt_params, :force, true)
```
extract to private function to make sure we use the same defaults everywhere

jlgeering replied:
or default value map?

Commit:
put defaults in a constant

jlgeering commented:
LGTM
